sdks/java/core/src/main/java/org/apache/beam/sdk/util/state/InMemoryTimerInternals.java

```diff
@@ -0,0 +1,234 @@
+    checkNotNull(newOutputWatermark);
+    if (newOutputWatermark.isAfter(inputWatermarkTime)) {
+      WindowTracing.trace(
+          "TestTimerInternals.advanceOutputWatermark: clipping output watermark from {} to {}",
+          newOutputWatermark,
+          inputWatermarkTime);
+      newOutputWatermark = inputWatermarkTime;
+    }
+    checkState(
+        outputWatermarkTime == null || !newOutputWatermark.isBefore(outputWatermarkTime),
+        "Cannot move output watermark time backwards from %s to %s",
+        outputWatermarkTime,
+        newOutputWatermark);
+    WindowTracing.trace(
+        "TestTimerInternals.advanceOutputWatermark: from {} to {}",
+        outputWatermarkTime,
+        newOutputWatermark);
+    outputWatermarkTime = newOutputWatermark;
+  }
+
+  /** Advances processing time to the given value and fires processing-time timers accordingly. */
+  public void advanceProcessingTime(
+      @Nullable TimerCallback timerCallback, Instant newProcessingTime) throws Exception {
+    checkState(
+        !newProcessingTime.isBefore(processingTime),
+        "Cannot move processing time backwards from %s to %s",
+        processingTime,
+        newProcessingTime);
+    WindowTracing.trace(
+        "TestTimerInternals.advanceProcessingTime: from {} to {}",
+        processingTime,
+        newProcessingTime);
+    processingTime = newProcessingTime;
+    advanceAndFire(timerCallback, newProcessingTime, TimeDomain.PROCESSING_TIME);
+  }
+
+  /**
+   * Advances synchronized processing time to the given value and fires processing-time timers
+   * accordingly.
+   */
+  public void advanceSynchronizedProcessingTime(
+      @Nullable TimerCallback timerCallback, Instant newSynchronizedProcessingTime)
+      throws Exception {
+    checkState(
+        !newSynchronizedProcessingTime.isBefore(synchronizedProcessingTime),
+        "Cannot move processing time backwards from %s to %s",
+        processingTime,
+        newSynchronizedProcessingTime);
+    WindowTracing.trace(
+        "TestTimerInternals.advanceProcessingTime: from {} to {}",
+        synchronizedProcessingTime,
+        newSynchronizedProcessingTime);
+    synchronizedProcessingTime = newSynchronizedProcessingTime;
+    advanceAndFire(
+        timerCallback, newSynchronizedProcessingTime, TimeDomain.SYNCHRONIZED_PROCESSING_TIME);
+  }
+
+  private void advanceAndFire(
+      @Nullable TimerCallback timerCallback, Instant currentTime, TimeDomain domain)
+      throws Exception {
+    PriorityQueue<TimerData> queue = queue(domain);
+    while (!queue.isEmpty() && currentTime.isAfter(queue.peek().getTimestamp())) {
+      TimerData timer = queue.peek();
+      WindowTracing.trace(
+          "InMemoryTimerInternals.advanceAndFire: firing {} at {}", timer, currentTime);
+      // Remove before firing, so that if the callback adds another identical
+      // timer we don't remove it.
+      queue.remove();
+      if (timerCallback != null) {
```

Comment:
I was using null as "no-op callback" (for users who only care about the times of timers, but not about doing anything with them), but now replaced it with an explicit NO_OP callback.